### PR TITLE
Feature/CVSB-10283 -  VIEW test result

### PR DESCRIPTION
**Pull request by irina005**
Ticket:
https://jira.dvsacloud.uk/browse/CVSB-10283
Old pr:
https://github.com/dvsa/cvs-app-vtm/pull/125/files

### CODE REVIEW

File: src/.prettierrc [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (alleypa) [inline src/.prettierrc:1]: please remove from here

File: src/app/pipes/DefaultNullOrEmptyPipe.spec.ts [hunk 2]
```diff
@@ -31,29 +31,15 @@ describe('DefaultNullOrEmptyPipe', () => {
...
-    expect(checkDate.toISOString()).toEqual(stringToCheck);
-    expect(res).toEqual('14/02/2019');
-  });
-
   it('a string of 2020-06-24 should be displayed in proper format', () => {
-    const stringToCheck = '2020-06-24';
+    const stringToCheck = '24/06/2020';
     const res = pipe.transform(stringToCheck);
     expect(res).toEqual('24/06/2020');
   });
```
Comment (alleypa) [inline src/app/pipes/DefaultNullOrEmptyPipe.spec.ts:45]: this test is no longer valid. Use `it should return expected string`. Meaning the parsed string has not been processed. You  don't need to use a date string for test

File: src/app/pipes/display-by-delimiter.spec.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { DisplayByDelimiter } from '@app/pipes/display-by-delimiter';
+
+describe('DisplayByDelimiter', () => {
+  const pipe = new DisplayByDelimiter();
+
+  it('create an instance', () => {
+    expect(pipe).toBeTruthy();
+  });
+
+  it('should return Yes if value is boolean', () => {
```
Comment (alleypa) [inline src/app/pipes/display-by-delimiter.spec.ts:10]: `it should return the expected concatenation`. You also need 2 more tests.
- a test with concatenation and having no delimiter
- a test that takes empty string and does nothing

File: src/app/shared/error-summary/__snapshots__/error-summary.component.spec.ts.snap [hunk 4]
```diff
@@ -0,0 +1,56 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (alleypa) [inline src/app/shared/error-summary/__snapshots__/error-summary.component.spec.ts.snap:1]: what has changed here. this should not be part of your commit

File: src/assets/config/config.dev.json [hunk 5]
```diff
@@ -18,7 +17,7 @@
     "clientId": "",
     "tenant": "",
     "redirectUri": "http://localhost:4200",
-    "cacheLocation": "localStorage",
+    "cacheLocation": "",
```
Comment (alleypa) [inline src/assets/config/config.dev.json:20]: undo this

File: src/app/pipes/DefaultNullOrEmptyPipe.spec.ts [hunk 6]
```diff
@@ -31,30 +31,16 @@ describe('DefaultNullOrEmptyPipe', () => {
...
-  });
-
-  it('a string of 2020-06-24 should be displayed in proper format', () => {
-    const stringToCheck = '2020-06-24';
-    const res = pipe.transform(stringToCheck);
-    expect(res).toEqual('24/06/2020');
+  it('should return expected string', () => {
+    const expectedRes = 'testing string';
+    const res = pipe.transform(expectedRes);
+    expect(res).toEqual('testing string');
```
Comment (JonLazariniDD) [inline src/app/pipes/DefaultNullOrEmptyPipe.spec.ts:44]: I think we need one more unit check that tests whether we return - for the else condition, also something minor but i think we could save a few lines if use a 'classic' if else, i don't think we need the switch case statement anymore

File: src/app/pipes/capitalize-string.spec.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+  const pipe = new CapitalizeString();
+  const stringToCapitalize = 'test';
+
+  it('create an instance', () => {
+    expect(pipe).toBeTruthy();
+  });
+
+  it('should return capitalized word', () => {
+    const res = pipe.transform(stringToCapitalize);
+    expect(res).toEqual('Test');
```
Comment (JonLazariniDD) [inline src/app/pipes/capitalize-string.spec.ts:13]: Same for this one i think we could do one more test that checks the else of ternary operator, as it seems that if we give something 'empty' to the pipe, it returns ''

File: src/app/pipes/display-by-delimiter.spec.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+  it('should return strings concatenated with no delimiter', () => {
+    const stringsConcatenated = 'test1test2';
+    const res = pipe.transform(stringsToConcatenate, '');
+    expect(res).toEqual(stringsConcatenated);
+  });
+
+  it('should not concatenate if empty string is given', () => {
+    const stringsConcatenated = '';
+    const res = pipe.transform([''], '');
```
Comment (JonLazariniDD) [inline src/app/pipes/display-by-delimiter.spec.ts:25]: 👍

File: src/app/shared/libraries/accordion-item/accordion-item.component.html [hunk 9]
```diff
@@ -0,0 +1,15 @@
+
+<div class="govuk-accordion__section">
+  <div class="govuk-accordion__section-header">
+    <h2 class="govuk-accordion__section-heading">
+      <span class="govuk-link govuk-!-font-size-36">{{title}}</span>
+      <span class="govuk-accordion__section-button" id="accordion-default-heading"></span>
```
Comment (JonLazariniDD) [inline src/app/shared/libraries/accordion-item/accordion-item.component.html:6]: I think it might need to be test-accordion-default-heading if we have an AC for that so that it can be picked up by the automation team.

File: src/app/shared/libraries/accordion-item/accordion-item.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(AccordionItemComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
```
Comment (JonLazariniDD) [inline src/app/shared/libraries/accordion-item/accordion-item.component.spec.ts:23]: a snapshot would be good to make sure we have the DOM rendered.

File: src/app/shared/libraries/accordion/accordion.component.html [hunk 11]
```diff
@@ -0,0 +1,6 @@
+<div class="govuk-accordion" data-module="govuk-accordion" id="accordion-default">
```
Comment (JonLazariniDD) [inline src/app/shared/libraries/accordion/accordion.component.html:1]: might need to prepend this id with test- in case this has to be used by the automation team

File: src/app/shared/libraries/accordion/accordion.component.spec.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(AccordionComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
```
Comment (JonLazariniDD) [inline src/app/shared/libraries/accordion/accordion.component.spec.ts:23]: one more snapshot test will be good also I think.

File: src/app/store/actions/VehicleTestResultModel.actions.spec.ts [hunk 13]
```diff
@@ -1,36 +1,15 @@
...
-  noOfAxles: 2,
-  vehicleType: 'none',
-  countryOfRegistration: 'Ro',
-  preparerId: '231',
-  odometerReading: 123,
-  vehicleConfiguration: 'none',
-  testStationType: 'none',
-  reasonForCancellation: 'none',
-  testerName: '213',
+const testResultModelExample = {
```
Comment (JonLazariniDD) [inline src/app/store/actions/VehicleTestResultModel.actions.spec.ts:6]: As the code in place was prior to ADR, i have added a technical task ticket to refactor the state names for test result as well as we discussed and agreed yesterday.
Please let me know if you don't agree.

https://jira.dvsacloud.uk/browse/CVSB-15127

File: src/app/store/selectors/VehicleTestResultModel.selectors.ts [hunk 14]
```diff
@@ -9,3 +11,26 @@ export const selectSelectedVehicleTestResultModel = createSelector(
...
   (state: IVehicleTestResultModelState) => state.vehicleTestResultModel
 );
+
+export const selectTestTypeById = (id: string) => createSelector(
+  selectFeature,
+  (state: IVehicleTestResultModelState) => {
+    let testType: TestType;
+    let testRecord: TestResultModel;
+
+    if (state.vehicleTestResultModel) {
```
Comment (JonLazariniDD) [inline src/app/store/selectors/VehicleTestResultModel.selectors.ts:21]: This will need a unit test for that.
We can extract that into a separate function if it makes it hard to unit test that we then use in this selector.
It seems like it is taking a vehicleTestResultModel as input and testType.

File: src/app/technical-record/adr/adr-details/adr-details.component.html [hunk 15]
```diff
@@ -13,7 +13,7 @@ <h2 class="govuk-heading-m">ADR details</h2>
     <div class="govuk-summary-list__row">
       <dt class="govuk-summary-list__key">Date processed</dt>
       <dd id="test-ADR-approvalDate" class="govuk-summary-list__value">
-        {{ adrDetails.vehicleDetails?.approvalDate | DefaultNullOrEmpty }}
+        {{ adrDetails.vehicleDetails?.approvalDate ? (adrDetails.vehicleDetails?.approvalDate | date: 'dd/MM/yyyy') : '-' }}
```
Comment (JonLazariniDD) [inline src/app/technical-record/adr/adr-details/adr-details.component.html:16]: Can we use a pipe instead for this please? If easy to do, I believe we could chain Pipes.
It would be best avoiding inline Js in the markup

File: src/app/technical-record/test-history/test-history.component.html [hunk 16]
```diff
@@ -7,20 +7,20 @@
...
   <tbody class="govuk-table__body" *ngIf="testResultJson; else noTestHistory">
   <ng-container *ngFor="let testRes of (testResultJson | SortByDatePipe :'testEndTimestamp'); let i = index">
     <tr class="govuk-table__row" *ngFor="let testType of (testRes.testTypes | SortByDatePipe :'testTypeEndTimestamp'); let ii = index">
       <td id="{{ 'test-testCode-tRes-' + i + '-tType-' + ii }}" class="govuk-table__cell">
-        {{ techRecHelpers.capitalizeWord(testType.testCode | DefaultNullOrEmpty) }}
+        {{ (testType.testCode | CapitalizeString) | DefaultNullOrEmpty }}
       </td>
       <td id="{{ 'test-testTypeEndTimestamp-tRes-' + i + '-tType-' + ii }}" class="govuk-table__cell">
-        {{ testType.testTypeEndTimestamp | DefaultNullOrEmpty }}
+        {{ testType.testTypeEndTimestamp ? (testType.testTypeEndTimestamp | date: 'dd/MM/yyyy') : '-' }}
```
Comment (JonLazariniDD) [inline src/app/technical-record/test-history/test-history.component.html:20]: Same i think we could do with a Pipe maybe since it is reused multiple times

File: src/app/technical-record/vehicle-summary/vehicle-summary.component.html [hunk 17]
```diff
@@ -22,7 +22,7 @@
       Date of first registration
     </dt>
     <dd id="test-regnDate" class="govuk-summary-list__value">
-      {{ activeRecord.regnDate | DefaultNullOrEmpty }}
+      {{ activeRecord.regnDate ? (activeRecord.regnDate | date: 'dd/MM/yyyy') : '-' }}
```
Comment (JonLazariniDD) [inline src/app/technical-record/vehicle-summary/vehicle-summary.component.html:25]: Same for those

File: src/app/test-record/defects/defects.component.spec.ts [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+  beforeEach(() => {
+    fixture = TestBed.createComponent(DefectsComponent);
+    component = fixture.componentInstance;
+    component.testType = testType;
+    component.testRecord = testRecord;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
```
Comment (JonLazariniDD) [inline src/app/test-record/defects/defects.component.spec.ts:29]: One snapshot would be good if possible also please.

File: src/app/test-record/emission-details/emission-details.component.spec.ts [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(EmissionDetailsComponent);
+    component = fixture.componentInstance;
+    component.testType = testType;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
```
Comment (JonLazariniDD) [inline src/app/test-record/emission-details/emission-details.component.spec.ts:27]: snapshot if possible would be good please.

File: src/app/test-record/notes/notes.component.spec.ts [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(NotesComponent);
+    component = fixture.componentInstance;
+    component.testType = testType;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
```
Comment (JonLazariniDD) [inline src/app/test-record/notes/notes.component.spec.ts:26]: Snapshot for this would be good if possible please

File: src/app/test-record/seatbelt-installation-check/seatbelt-installation-check.component.html [hunk 21]
```diff
@@ -0,0 +1,26 @@
+<dl class="govuk-summary-list">
+  <div class="govuk-summary-list__row">
+    <dt class="govuk-summary-list__key">
+      Carried out during test
+    </dt>
+    <dd class="govuk-summary-list__value" id="test-seatbeltInstallationCheckDate">
+      {{ testType.seatbeltInstallationCheckDate ? (testType.seatbeltInstallationCheckDate| date: 'dd/MM/yyyy') : '-' }}
```
Comment (JonLazariniDD) [inline src/app/test-record/seatbelt-installation-check/seatbelt-installation-check.component.html:7]: Same for this, i think we could do with a Pipe if possible.

File: src/app/test-record/seatbelt-installation-check/seatbelt-installation-check.component.spec.ts [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(SeatbeltInstallationCheckComponent);
+    component = fixture.componentInstance;
+    component.testType = testType;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
```
Comment (JonLazariniDD) [inline src/app/test-record/seatbelt-installation-check/seatbelt-installation-check.component.spec.ts:26]: a snapshot would be good also please

File: src/app/test-record/test-history/test-history.component.spec.ts [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(TestHistoryComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
```
Comment (JonLazariniDD) [inline src/app/test-record/test-history/test-history.component.spec.ts:23]: snapshot here would be good also

File: src/app/test-record/test-record.component.html [hunk 24]
```diff
@@ -1,4 +1,76 @@
-<p>
-  <span>JSON response:</span>
-  {{ testType | json }}
-</p>
+<div class="govuk-width-container">
+  <a class="govuk-back-link" vtmBackButton>Back</a>
```
Comment (JonLazariniDD) [inline src/app/test-record/test-record.component.html:2]: probably one for later but I think this should be part of the shell component which should contain
the navigation, header, container (showing tech-record and test-result sections)
I know this was in place prior to ADR i believe so this will need some attention but will require some time to carefully move the right pieces around.

File: src/app/test-record/test-record.component.spec.ts [hunk 25]
```diff
@@ -43,11 +45,16 @@ describe('TestRecordComponent', () => {
...
+    component.testType = testType;
+    component.testRecord = testRecord;
+    component.seatBeltApplicable = ['11'];
+    component.emissionDetailsApplicable = ['22'];
+    component.defectsApplicable = ['33'];
     fixture.detectChanges();
   });
 
   it('should create', () => {
     expect(component).toBeTruthy();
```
Comment (JonLazariniDD) [inline src/app/test-record/test-record.component.spec.ts:57]: would be nice to add a snapshot and give a bit more meaningful description like we have done.

File: src/app/test-record/test-record.component.ts [hunk 26]
```diff
@@ -1,43 +1,44 @@
...
-          if (res.testNumber === tTypeNumber) {
-            tType = res;
-            return res;
-          }
-        });
-      });
-      this.testType = tType;
-    });
+  goBack() {
+    window.history.back();
```
Comment (JonLazariniDD) [inline src/app/test-record/test-record.component.ts:42]: Need to add a unit test for that as well if possible we can spy on window.history.back, stub it and make sure it (mock) is called when clicked.

File: src/app/test-record/test-record.container.spec.ts [hunk 27]
```diff
@@ -0,0 +1,59 @@
...
+    container.testRecordObservable$ = of(testRecord);
+    container.testTypeNumber = 'W01A34247';
+    container.seatBeltApplicable = ['11'];
+    container.emissionDetailsApplicable = ['22'];
+    container.defectsApplicable = ['33'];
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(container).toBeTruthy();
```
Comment (JonLazariniDD) [inline src/app/test-record/test-record.container.spec.ts:57]: Need to add a snapshot for that as well I think.

File: src/app/test-record/vehicle/vehicle.component.html [hunk 28]
```diff
@@ -0,0 +1,66 @@
+<dl *ngIf="testRecord" class="govuk-summary-list">
+  <div class="govuk-summary-list__row">
+    <dt class="govuk-summary-list__key">
+      Vehicle identification number (VIN)
+    </dt>
+    <dd class="govuk-summary-list__value" id="test-vin">
+      {{ testRecord.vin | DefaultNullOrEmpty }}
+    </dd>
+  </div>
+  <div *ngIf="testRecord.vehicleType !== 'trl'" class="govuk-summary-list__row">
```
Comment (JonLazariniDD) [inline src/app/test-record/vehicle/vehicle.component.html:10]: This is one for later but we will need some mapping exercise for that like we did for tech-record

(FYI @alleypa @m-ujaffer @rspasici )

File: src/app/technical-record/adr/adr-details/adr-details.component.html (hunk 15 above)
Comment (alleypa) [reply src/app/technical-record/adr/adr-details/adr-details.component.html:16]: Hi @JonLazariniDD  i think this if fine here. The nullable operator `?.` Serves as a shortcut in this case. Having another will be over kill

File: src/app/store/selectors/VehicleTestResultModel.selectors.ts (hunk 14 above)
Comment (irina005) [reply src/app/store/selectors/VehicleTestResultModel.selectors.ts:21]: This selector has been simplified in the implementation of 10278

File: src/app/test-record/vehicle/vehicle.component.spec.ts [hunk 29]
```diff
@@ -0,0 +1,28 @@
...
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(VehicleComponent);
+    component = fixture.componentInstance;
+    component.testRecord = testRecord;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
```
Comment (JonLazariniDD) [inline src/app/test-record/vehicle/vehicle.component.spec.ts:26]: I think we need to add one snapshot for that as well please.

File: src/app/test-record/visit/visit.component.spec.ts [hunk 30]
```diff
@@ -0,0 +1,29 @@
...
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(VisitComponent);
+    component = fixture.componentInstance;
+    component.testRecord = testRecord;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
```
Comment (JonLazariniDD) [inline src/app/test-record/visit/visit.component.spec.ts:27]: snapshot for that would be good please.

File: src/app/technical-record/adr/tank-inpections/__snapshots__/tank-inpections.component.spec.ts.snap [hunk 31]
```diff
@@ -74,7 +74,7 @@ exports[`TankInpectionsComponent should create view only with populated data 1`]
                 class="govuk-table__cell"
                 id="test-tc2IntermediateExpiryDate"
               >
-                 18/03/2020 
+                 2020-03-18
```
Comment (alleypa) [inline src/app/technical-record/adr/tank-inpections/__snapshots__/tank-inpections.component.spec.ts.snap:77]: this is actually broken. it should be `18/03/2020` in the result

Comment (irina005) [reply src/app/technical-record/adr/tank-inpections/__snapshots__/tank-inpections.component.spec.ts.snap:77]: I missed this one, apologies

File: src/app/test-record/vehicle/vehicle.component.spec.ts (hunk 29 above)
Comment (irina005) [reply src/app/test-record/vehicle/vehicle.component.spec.ts:26]: Added in 10278

File: src/app/test-record/visit/visit.component.spec.ts (hunk 30 above)
Comment (irina005) [reply src/app/test-record/visit/visit.component.spec.ts:27]: Added in 10278

File: src/app/test-record/test-section/test-section.component.spec.ts [hunk 32]
```diff
@@ -0,0 +1,39 @@
...
+    component = fixture.componentInstance;
+    component.testType = testType;
+    component.testRecord = testRecord;
+    component.applicableTestTypeIds1 = { 43: 'test' };
+    component.applicableTestTypeIds2 = { 43: 'test' };
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
```
[the remark this answers is not in the document]
Comment (irina005) [reply src/app/test-record/test-section/test-section.component.spec.ts:37]: Added in 10278

File: src/app/test-record/test-history/test-history.component.scss [hunk 33]
```diff
@@ -0,0 +1,7 @@
+.govuk-summary-list {
```
Comment (alleypa) [inline src/app/test-record/test-history/test-history.component.scss:1]: we should use you new css class for this
